app/queries/user_analytics_query.rb

```diff
@@ -3,33 +3,34 @@ class UserAnalyticsQuery
   include MonthHelper
   include PeriodHelper
 
-  attr_reader :current_facility, :days_ago, :months_ago
```

Comment:
Why did we axe the attr_readers? Can we reinstate them?

Reply:
I axed them because I have become disillusioned.

I have become disillusioned by the best-practice OOisms from the Sandi Metzes and the Uncle Bobs of the world.

Lot of the semantic best practices of OO, I feel, come as a protection against dubious programmers and programming. As I understand it, the common reason for sticking this piece of *data* in an `attr_reader` is to make it *behavior*. This ostensibly allows for two things:

1. You can change the meaning of this `attr_reader` by re-defining it later, as a proper public method, and modifying its *behavior* without changing any code that depends on it. For example: `days_ago` can later become: `def days_ago; @days_ago - 1; end` and have it be propagated everywhere.
2. You deal with the piece of *data* in an effectively immutable way.

I find this form of I̧̳͇̦n͍̞̥̙̹fo̟͕r̡̧̗͇ma̡̫͇͔t̢̙̣͚͟i̞̥̺̥on̼ H̫͓͍̰įd͍͉̫̻i̡͕n̞g͇̬͇̠̘ a local-maxima optimization.

1., has two problems in my head,

a. **Behavior**: This conflates data with behavior. There's now no distinction between them. Classes are nothing but data + behavior, and we've made all data behavior just for refactoring ease for the future, that RubyMine can do in 5 seconds.
b. **Visibility**: This class now needlessly exposes pieces of data as behavior (get-ers) when the interface shouldn't really care about reading `days_ago` off of the object at all, further complicating what the public API of this class is.

2., is faking immutability and a protection against dubious programming. If we're worried about typos causing us state issues, we should invest in languages that are typed and support immutability as first-class citizens.

I see another general issue with this pattern (an extension of 1a.), that is of defining the boundaries of what should be considered data. Data is data. It's never anything else. As soon as it becomes non-static, it's a transformation on the data. It's important to know what is clearly static and what isn't, when looking at a class, imo.

Similarly (an extension of 1b.), it's important to clearly know what the public API of the class is by just looking at. It is the public methods + `attr_*`.

**Caveats**

* None of this is a stone-tablet. All of it is just my opinion. I'm happy to change it if this is a pattern we'd like to adopt.
* Some of the problems of 1b. can be solved by a pattern that I've [used in the past](https://github.com/simpledotorg/simple-server/blob/master/app/services/sms_notification_service.rb#L27). This allows the reader to be non-public, but still act as *behavior* and be immutable. This is not always feasible though since the boundaries of access become unclear when you start using private `attr_readers` in public methods of the class. I'm tentatively open to adopting this pattern though, if we like it.
* This entire comment is intentionally written in a lightly-frustrated tone. I'm not actually salty about any of this 👾

Reply:
I definitely do not feel as strongly about this one way or the other 😅 Let's axe em! Be sure to catch these in my PRs as well 👍